src/include/migraphx/op/topk.hpp

```diff
@@ -0,0 +1,234 @@
+#ifndef MIGRAPHX_GUARD_OPERATORS_GATHER_HPP
+#define MIGRAPHX_GUARD_OPERATORS_GATHER_HPP
+
+#include <algorithm>
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/value.hpp>
+#include <migraphx/op/normalize_attribute.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/par_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct topk
+{
+    int64_t k;
+    int64_t axis = 0;
+    bool largest = true;
+    bool sorted  = true;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.k, "k"),
+                    f(self.axis, "axis"),
+                    f(self.largest, "largest"),
+                    f(self.sorted, "sorted"));
+    }
+
+    value attributes() const
+    {
+        value normalize;
+        normalize["axis"] = value::array{normalize_attribute::include_min};
+        return {{"normalize_axes", normalize}};
+    }
+
+    std::string name() const { return "topk"; }
+
+    shape normalize_compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this}.has(1).standard();
+        auto lens = inputs.at(0).lens();
+        auto type = inputs.at(0).type();
+
+        lens[axis] = k;
+
+        shape s_val{type, lens};
+        shape s_ind{shape::int64_type, lens};
+
+        return shape({s_val, s_ind});
+    }
+
+    struct compare_op
+    {
+        bool larger;
+        template <class T, class U>
+        auto operator()(T x, U y) const
+        {
+            return (larger ? (x < y) : (x > y));
+        }
+    };
```

Comment:
The compare_op should be passed in as a template parameter so we can pass `std::less<>{}` or `std::greater<>{}`. The same would apply to the GPU implementation, as we are doing an extra comparison everytime we compare elements.

Reply:
fixed in both ref and gpu implementation.